File: modules/mastodon/repository/src/rust/conversion/status.rs
/*
 * Copyright 2025 wcaokaze
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use mastodon_entity::instance::Instance;
use mastodon_entity::status::{Status, StatusHashtag, StatusMention};
use panoptiqon::cache::Cache;
use crate::cache;

use mastodon_webapi::entity::status::{
   Status as ApiStatus,
   StatusMention as ApiStatusMention,
   StatusTag as ApiStatusTag,
};

#[cfg(feature = "jvm")]
use jni::JNIEnv;

pub fn from_api(
   #[cfg(feature = "jvm")] env: &mut JNIEnv,
   instance: Cache<Instance>,
   entity: ApiStatus,
   account_cache_repository: &mut cache::account::Repository,
   status_cache_repository: &mut cache::status::StatusRepository,
   no_credential_status_cache_repository:
      &mut cache::status::NoCredentialStatusRepository,
   no_credential_poll_repository: &mut cache::poll::NoCredentialPollRepository,
) -> anyhow::Result<Status> {
   use anyhow::Context;
   use chrono::DateTime;
   use isolang::Language;
   use mastodon_entity::status::{
      NoCredentialStatus, StatusId, StatusLocalId, StatusVisibility,
   };
   use mastodon_entity::account::{AccountId, AccountLocalId};
   use crate::conversion;

   let ApiStatus {
      id, uri, created_at, account, content, visibility, sensitive, spoiler_text,
      media_attachments, application, mentions, tags, emojis, reblogs_count,
      favourites_count, replies_count, url, in_reply_to_id, in_reply_to_account_id,
      reblog, poll, card, language, text, edited_at, favourited, reblogged, muted,
      bookmarked, pinned, filtered,
   } = entity;

   let id = StatusId {
      instance_url: instance.get().url.clone(),
      local: StatusLocalId(id.context("No status id")?)
   };

   let boosted_status = reblog
      .and_then(|reblog|
         from_api(
            #[cfg(feature = "jvm")] env,
            instance.clone(),
            *reblog,
            account_cache_repository,
            status_cache_repository,
            no_credential_status_cache_repository,
            no_credential_poll_repository
         ).ok()
      )
      .map(|boosted| status_cache_repository.save(boosted));

   let poll = poll.and_then(|poll|
      conversion::poll::from_api(
         instance.clone(), poll, no_credential_poll_repository
      ).ok()
   );

   let no_credential = NoCredentialStatus {
      id: id.clone(),
      uri,
      created_time: created_at
         .and_then(|time| DateTime::parse_from_rfc3339(&time).ok())
         .map(|time| time.to_utc()),
      account: account
         .and_then(|acc|
            conversion::account::from_api(
               #[cfg(feature = "jvm")] env,
               instance.clone(),
               acc,
               account_cache_repository
            ).ok()
         )
         .map(|acc| account_cache_repository.save(acc)),
      content,
      visibility: visibility.map(StatusVisibility),
      is_sensitive: sensitive,
      spoiler_text,
      media_attachments: media_attachments.unwrap_or(vec![]).into_iter()
         .flat_map(|media|
            conversion::media_attachment::from_api(media)
         )
         .collect(),
      application: application.and_then(|app|
         conversion::application::from_api(app, instance.clone()).ok()
      ),
      mentions: mentions.unwrap_or(vec![]).into_iter()
         .flat_map(|mnt| mention_from_api(instance.clone(), mnt))
         .collect(),
      hashtags: tags.unwrap_or(vec![]).into_iter()
         .flat_map(|tag| tag_from_api(tag))
         .collect(),
      emojis: emojis.unwrap_or(vec![]).into_iter()
         .flat_map(|emj| conversion::custom_emoji::from_api(instance.clone(), emj))
         .collect(),
      boost_count: reblogs_count,
      favorite_count: favourites_count,
      reply_count: replies_count,
      url: url.and_then(|url| url.parse().ok()),
      replied_status_id: in_reply_to_id.map(|id|
         StatusId {
            instance_url: instance.get().url.clone(),
            local: StatusLocalId(id)
         }
      ),
      replied_account_id: in_reply_to_account_id.map(|id|
         AccountId {
            instance_url: instance.get().url.clone(),
            local: AccountLocalId(id)
         }
      ),
      boosted_status: boosted_status.as_ref().map(|s| s.get().no_credential.clone()),
      poll: poll.as_ref().map(|p| p.no_credential.clone()),
      card: card.and_then(|card|
         conversion::preview_card::from_api(
            #[cfg(feature = "jvm")] env,
            instance.clone(),
            card,
            account_cache_repository
         ).ok()
      ),
      language: language.and_then(|code| Language::from_639_1(&code)),
      text,
      edited_time: edited_at
         .and_then(|time| DateTime::parse_from_rfc3339(&time).ok())
         .map(|time| time.to_utc()),
   };

   let no_credential = no_credential_status_cache_repository.save(no_credential);

   let status = Status {
      id: id.clone(),
      no_credential,
      boosted_status,
      poll,
      is_favorited: favourited,
      is_boosted: reblogged,
      is_muted: muted,
      is_bookmarked: bookmarked,
      is_pinned: pinned,
      filter_results: filtered.unwrap_or(vec![]).into_iter()
         .flat_map(|fr|
            conversion::filter::filter_result_from_api(
               instance.get().url.clone(), fr
            )
         )
         .collect(),
   };

   Ok(status)
}

pub fn mention_from_api(
   instance: Cache<Instance>,
   entity: ApiStatusMention
) -> anyhow::Result<StatusMention> {
   use mastodon_entity::account::{AccountId, AccountLocalId};

   let ApiStatusMention { id, username, url, acct } = entity;

   let status_mention = StatusMention {
      mentioned_account_id: id.map(|id| AccountId {
         instance_url: instance.get().url.clone(),
         local: AccountLocalId(id)
      }),
      mentioned_account_username: username,
      mentioned_account_url: url.and_then(|url| url.parse().ok()),
      mentioned_account_acct: acct,
   };

   Ok(status_mention)
}

pub fn tag_from_api(
   entity: ApiStatusTag
) -> anyhow::Result<StatusHashtag> {
   let ApiStatusTag { name, url } = entity;

   let status_tag = StatusHashtag {
      name,
      url: url.and_then(|url| url.parse().ok()),
   };

   Ok(status_tag)
}

File: modules/mastodon/entity/src/commonMain/kotlin/com/wcaokaze/probosqis/mastodon/entity/Status.kt
/*
 * Copyright 2025 wcaokaze
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wcaokaze.probosqis.mastodon.entity

import com.wcaokaze.probosqis.ext.kotlin.Url
import com.wcaokaze.probosqis.panoptiqon.Cache
import kotlinx.datetime.Instant

data class Status(
   val id: Id,
   val noCredential: Cache<NoCredential>,
   val boostedStatus: Cache<Status>?,
   val poll: Poll?,
   val isFavorited: Boolean?,
   val isBoosted: Boolean?,
   val isMuted: Boolean?,
   val isBookmarked: Boolean?,
   val isPinned: Boolean?,
   val filterResults: List<FilterResult>,
) {
   constructor(
      rawInstanceUrl: String,
      rawLocalId: String,
      noCredential: Cache<NoCredential>,
      boostedStatus: Cache<Status>?,
      poll: Poll?,
      isFavorited: Boolean?,
      isBoosted: Boolean?,
      isMuted: Boolean?,
      isBookmarked: Boolean?,
      isPinned: Boolean?,
      filterResults: List<FilterResult>,
   ) : this(
      Id(Url(rawInstanceUrl), LocalId(rawLocalId)),
      noCredential,
      boostedStatus,
      poll,
      isFavorited,
      isBoosted,
      isMuted,
      isBookmarked,
      isPinned,
      filterResults,
   )

   val rawInstanceUrl: String
      get() = id.instanceUrl.raw

   val rawLocalId: String
      get() = id.local.raw

   data class Id(
      val instanceUrl: Url,
      val local: LocalId
   ) {
      constructor(
         rawInstanceUrl: String,
         rawLocalId: String,
         @Suppress("UNUSED_PARAMETER")
         dummy: Unit?
      ) : this(
         Url(rawInstanceUrl),
         LocalId(rawLocalId)
      )

      val rawInstanceUrl: String
         get() = instanceUrl.raw

      val rawLocalId: String
         get() = local.raw

      val dummy: Unit?
         get() = null
   }

   @JvmInline
   value class LocalId(val raw: String)

   data class NoCredential(
      val id: Id,
      val uri: String?,
      val createdTime: Instant?,
      val account: Cache<Account>?,
      val content: String?,
      val visibility: Visibility?,
      val isSensitive: Boolean?,
      val spoilerText: String?,
      val mediaAttachments: List<MediaAttachment>,
      val application: Application?,
      val mentions: List<Mention>,
      val hashtags: List<Hashtag>,
      val emojis: List<CustomEmoji>,
      val boostCount: Long?,
      val favoriteCount: Long?,
      val replyCount: Long?,
      val url: Url?,
      val repliedStatusId: Id?,
      val repliedAccountId: Account.Id?,
      val boostedStatus: Cache<NoCredential>?,
      val poll: Cache<Poll.NoCredential>?,
      val card: PreviewCard?,
      val language: String?,
      val text: String?,
      val editedTime: Instant?,
   ) {
      constructor(
         rawInstanceUrl: String,
         rawLocalId: String,
         uri: String?,
         createdTimeEpochMillis: Long?,
         account: Cache<Account>?,
         content: String?,
         rawVisibility: String?,
         isSensitive: Boolean?,
         spoilerText: String?,
         mediaAttachments: List<MediaAttachment>,
         application: Application?,
         mentions: List<Mention>,
         hashtags: List<Hashtag>,
         emojis: List<CustomEmoji>,
         boostCount: Long?,
         favoriteCount: Long?,
         replyCount: Long?,
         rawUrl: String?,
         rawRepliedStatusLocalId: String?,
         rawRepliedAccountLocalId: String?,
         boostedStatus: Cache<NoCredential>?,
         poll: Cache<Poll.NoCredential>?,
         card: PreviewCard?,
         language: String?,
         text: String?,
         editedTimeEpochMillis: Long?,
      ) : this(
         Id(Url(rawInstanceUrl), LocalId(rawLocalId)),
         uri,
         createdTimeEpochMillis?.let(Instant::fromEpochMilliseconds),
         account,
         content,
         rawVisibility?.let(::Visibility),
         isSensitive,
         spoilerText,
         mediaAttachments,
         application,
         mentions,
         hashtags,
         emojis,
         boostCount,
         favoriteCount,
         replyCount,
         rawUrl?.let(::Url),
         repliedStatusId = rawRepliedStatusLocalId?.let {
            Id(Url(rawInstanceUrl), LocalId(it))
         },
         repliedAccountId = rawRepliedAccountLocalId?.let {
            Account.Id(Url(rawInstanceUrl), Account.LocalId(it))
         },
         boostedStatus,
         poll,
         card,
         language,
         text,
         editedTimeEpochMillis?.let(Instant::fromEpochMilliseconds),
      )

      val rawInstanceUrl: String
         get() = id.instanceUrl.raw

      val rawLocalId: String
         get() = id.local.raw

      val createdTimeEpochMillis: Long?
         get() = createdTime?.toEpochMilliseconds()

      val rawVisibility: String?
         get() = visibility?.raw

      val rawUrl: String?
         get() = url?.raw

      val rawRepliedStatusLocalId: String?
         get() = repliedStatusId?.local?.raw

      val rawRepliedAccountLocalId: String?
         get() = repliedAccountId?.local?.value

      val editedTimeEpochMillis: Long?
         get() = editedTime?.toEpochMilliseconds()
   }

   @JvmInline
   value class Visibility(val raw: String) {
      companion object {
         val PUBLIC   = Visibility("public")
         val UNLISTED = Visibility("unlisted")
         val PRIVATE  = Visibility("private")
         val DIRECT   = Visibility("direct")
      }
   }

   data class Mention(
      val mentionedAccountId: Account.Id?,
      val mentionedAccountUsername: String?,
      val mentionedAccountUrl: Url?,
      val mentionedAccountAcct: String?,
   ) {
      constructor(
         rawInstanceUrl: String?,
         rawMentionedAccountLocalId: String?,
         mentionedAccountUsername: String?,
         rawMentionedAccountUrl: String?,
         mentionedAccountAcct: String?,
      ) : this(
         if (rawInstanceUrl == null || rawMentionedAccountLocalId == null) {
            null
         } else {
            Account.Id(
               Url(rawInstanceUrl),
               Account.LocalId(rawMentionedAccountLocalId)
            )
         },
         mentionedAccountUsername,
         rawMentionedAccountUrl?.let(::Url),
         mentionedAccountAcct,
      )

      val rawInstanceUrl: String?
         get() = mentionedAccountId?.instanceUrl?.raw

      val rawMentionedAccountLocalId: String?
         get() = mentionedAccountId?.local?.value

      val rawMentionedAccountUrl: String?
         get() = mentionedAccountUrl?.raw
   }

   data class Hashtag(
      val name: String?,
      val url: Url?,
   ) {
      constructor(
         name: String?,
         rawUrl: String?,
         @Suppress("UNUSED_PARAMETER")
         dummy: Unit?
      ) : this(
         name,
         rawUrl?.let(::Url),
      )

      val rawUrl: String?
         get() = url?.raw

      val dummy: Unit?
         get() = null
   }
}

tailrec fun Status.resolveBoostedStatus(): Status {
   if (boostedStatus == null) { return this }
   return boostedStatus.value.resolveBoostedStatus()
}

tailrec fun Status.NoCredential.resolveBoostedStatus(): Status.NoCredential {
   if (boostedStatus == null) { return this }
   return boostedStatus.value.resolveBoostedStatus()
}
